distributed/scheduler.py

```diff
@@ -5161,7 +5165,9 @@ async def restart(self, client=None, timeout=30):
 
         self.log_event([client, "all"], {"action": "restart", "client": client})
         start = time()
-        while time() < start + 10 and len(self.workers) < n_workers:
+        while time() < start + 10 and (
+            len(self.workers) < n_workers or initial_workers & self._get_worker_ids()
```

Comment:
Isn't a restarted worker going to have a new `server_id`? The ID is coming from the `Worker`, not the `Nanny` instance: https://github.com/dask/distributed/blob/d88c1d236da65f036af6ea05881c84f6b4c9fbb8/distributed/core.py#L341
https://github.com/dask/distributed/blob/d88c1d236da65f036af6ea05881c84f6b4c9fbb8/distributed/worker.py#L1113

So I don't know why we'd even expect those IDs to come back. Indeed, if they _did_ come back that would be highly concerning.

I think this might be reasonable (though obviously not tolerant to new workers joining during the restart):
```suggestion
            len(self.running) < len(nanny_workers)
```

Reply:
Is there a scenario where due to some race condition an old worker might somehow still be in `self.worker` or reconnecting before shutting down? I think that's what this is intended to catch.

Reply:
Yes, that's possible if we hit the `not all(resp == "OK" for resp in resps)` branch when telling Nanny workers to restart. If the nanny times out trying to kill the worker subprocess, that worker may still be connected.

https://github.com/dask/distributed/blob/f7f650154fea29978906c65dd0225415da56ed11/distributed/nanny.py#L479-L491

However, in that case, I think we'd want to ignore that worker in our count of workers we expect to come back. Since we didn't successfully shut it down, it's still around, and if/when it does shut down in the future, it's quite possible it won't come back. We can disregard the fact that we tried to restart that one; it's a lost cause.

Honestly, I wonder if we should call `remove_worker` on the workers that didn't respond `"OK"`. Their state is a bit undefined, but it's probably kinda broken. Maybe we don't want to talk to them anymore. If we remove them, the scheduler will reject them the next time they try to talk to it, so it's safe.

But if we don't do that, then I think `initial_workers & self._get_worker_ids()` is too strict. Because the only reason some `initial_workers` would still be connected is that their nannies failed to shut down. If they failed to shut down, then they may never restart, so we shouldn't wait for them.

I'd then be torn on `len(self.workers) < n_workers` vs `len(self.running) < n_workers`. It's quite possible these broken workers aren't running anymore, and never will be. So just checking `len(self.workers)` means they'll still be counted. OTOH, it's nice to wait for new workers to be fully running instead of just connected. We could do
```python
                if n_restart_failed := sum(resp != "OK" for resp in resps):
                    logger.error(
                        "Not all workers responded positively: %s",
                        resps,
                        exc_info=True,
                    )
                    n_workers -= n_restart_failed
```
Then checking `len(self.running) < n_workers` will ignore the workers we don't expect to restart. But I don't know if it's necessary.